src/storage/sftp/storage.c

```diff
@@ -37,6 +38,199 @@ struct StorageSftp
+            matchFailMsg = "mismatch in known hosts files: LIBSSH2_KNOWNHOST_CHECK_MISMATCH";
+            break;
+
+        default:
+            matchFailMsg = "experienced an unknown failure:";
+            break;
+    }
+
+    FUNCTION_TEST_RETURN_CONST(STRINGZ, matchFailMsg);
+}
+
+/***********************************************************************************************************************************
+Rewrite the user's known_hosts file with a new entry
+***********************************************************************************************************************************/
+static void
+storageSftpUpdateKnownHostsFile(StorageSftp *const this, const int hostKeyType, const String *const host, const char *const hostKey,
+        const size_t hostKeyLen)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(STORAGE_SFTP, this);
+        FUNCTION_LOG_PARAM(INT, hostKeyType);
+        FUNCTION_LOG_PARAM(STRING, host);
+        FUNCTION_LOG_PARAM(STRINGZ, hostKey);
+        FUNCTION_LOG_PARAM(SIZE, hostKeyLen);
+    FUNCTION_LOG_END();
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        LIBSSH2_KNOWNHOSTS *userKnownHostsList;
+        char *libSsh2ErrMsg;
+        int libSsh2ErrMsgLen;
+        int rc;
+
+        // Init a known host collection for the user's known_hosts file
+        const char *const userKnownHostsFile = strZ(strNewFmt("%s%s", strZ(userHome()), "/.ssh/known_hosts"));
+        userKnownHostsList = libssh2_knownhost_init(this->session);
+
+        LOG_WARN_FMT("host '%s' not found in known hosts files, attempting to add host to '%s'", strZ(host), userKnownHostsFile);
+
+        if (userKnownHostsList == NULL)
+        {
+            // Get the libssh2 error message and emit warning
+            rc = libssh2_session_last_error(this->session, &libSsh2ErrMsg, &libSsh2ErrMsgLen, 0);
+
+            LOG_WARN_FMT(
+                "libssh2_knownhost_init failed for '%s' for update: libssh2 errno [%d] %s", userKnownHostsFile, rc, libSsh2ErrMsg);
+        }
+        else
+        {
+            // Read the user's known_hosts file entries into the collection. libssh2_knownhost_readfile() returns the number of
+            // successfully loaded hosts or a negative value on error, an empty known hosts file will return 0.
+            if ((rc = libssh2_knownhost_readfile(userKnownHostsList, userKnownHostsFile, LIBSSH2_KNOWNHOST_FILE_OPENSSH)) < 0)
+            {
+                // Missing known_hosts file will return LIBSSH2_ERROR_FILE. Possibly issues other than missing may return this.
+                if (rc == LIBSSH2_ERROR_FILE)
+                {
+                    // If user's known_hosts file is non-existant, create an empty one for libssh2 to operate on.
+                    Storage *sshStorage =
+                        storagePosixNewP(
+                            strNewFmt("%s%s", strZ(userHome()), "/.ssh"), .modeFile = 0644, .modePath = 0700, .write = true);
```

Comment:
We should probably be doing modeFile = 0600, right?  That seems to be what ssh does when it creates a new .ssh and .ssh/known_hosts file.

Reply:
Updates in
https://github.com/pgbackrest/pgbackrest/pull/2125/commits/4fd017c80716d71a7ffff1f25379b9da196f7001
https://github.com/pgbackrest/pgbackrest/pull/2125/commits/6811405f150c8f1095c3da31d952f0ddebc58be1
https://github.com/pgbackrest/pgbackrest/pull/2125/commits/a3a82046feab0a915e0c116f867818b86ae42621

Reply:
Ah, alright, well, 0600 is probably fine if we're going to set a mode.  I'd be alright with letting umask apply too if that's easy to do.

Reply:
By default we set umask to 0: https://github.com/pgbackrest/pgbackrest/blob/release/2.47/src/config/load.c#L459

Reply:
Ok, so then setting mode to 0600 probably is best plan.  Resolving.